### PR TITLE
Fix: [common/utils/net.go bug](https://github.com/apache/dubbo-go/issues/182)

### PR DESCRIPTION
**Pull request by lazyfighter**
**What this PR does**:
#182 
**Which issue(s) this PR fixes**:
1. modify protocol/result.go member to private and provide two methods create result
2. remove common/utils/net.go filter private ip, and cache the localIp by global variable
3. use reflect.ValueOf(obj).MapKeys()  instead of MapRanges()  to compatible go1.11

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by hxmhlt** [comment]
Looks like travis-ci run failed. Pls check if you run go fmt ./... before commit.

File: cluster/cluster_impl/available_cluster_invoker.go [hunk 1]
```diff
@@ -18,14 +18,8 @@ limitations under the License.
 package cluster_impl
 
 import (
+	"errors"
 	"fmt"
-)
-
-import (
-	"github.com/pkg/errors"
-)
```
Comment (hxmhlt) [inline cluster/cluster_impl/available_cluster_invoker.go]: we need pkg/errors to output the package call stack

File: common/utils/net.go [hunk 2]
```diff
@@ -18,67 +18,35 @@
 package utils
 
 import (
+	"github.com/apache/dubbo-go/common/logger"
 	"net"
```
Comment (hxmhlt) [inline common/utils/net.go]: Pls format the import blocks. You can follow others files in dubbo-go.

File: common/utils/net.go [hunk 3]
```diff
@@ -18,67 +18,35 @@
...
 }
 
-// ref: https://stackoverflow.com/questions/23558425/how-do-i-get-the-local-ip-address-in-go
-func GetLocalIP() (string, error) {
-	ifs, err := net.Interfaces()
+func initLocalIp() {
+	addrs, err := net.InterfaceAddrs()
 	if err != nil {
-		return "", perrors.WithStack(err)
+		logger.Error("init local error", err)
```
Comment (AlexStocks) [inline common/utils/net.go]: add log prefix "[utils][initLocalIp]"

File: common/utils/net.go [hunk 4]
```diff
@@ -18,67 +18,35 @@
...
+		if ipNet, ok := addr.(*net.IPNet); ok && !ipNet.IP.IsLoopback() {
+			if ipNet.IP.To4() != nil {
+				localIp = ipNet.IP.String()
 			}
 		}
 	}
 
-	if ipAddr == nil {
-		return "", perrors.Errorf("can not get local IP")
+	if localIp == "" {
```
Comment (AlexStocks) [inline common/utils/net.go]: using "len(localIp) == 0" instead.

File: common/utils/net.go [hunk 5]
```diff
@@ -18,67 +18,35 @@
...
+		logger.Error("can not get local IP")
```
Comment (AlexStocks) [inline common/utils/net.go]: add log prefix "[utils][initLocalIp]"

File: protocol/jsonrpc/jsonrpc_invoker.go [hunk 6]
```diff
@@ -55,11 +55,14 @@ func (ji *JsonrpcInvoker) Invoke(invocation protocol.Invocation) protocol.Result
...
-		result.Rest = inv.Reply()
+
+	error := ji.client.Call(ctx, url, req, inv.Reply())
+	if error == nil {
+		result.SetResult(inv.Reply())
+	} else {
+		result.SetError(error)
 	}
-	logger.Debugf("result.Err: %v, result.Rest: %v", result.Err, result.Rest)
+	logger.Debugf("result.Err: %v, result.Rest: %v", result.Error(), result.Result())
```
Comment (AlexStocks) [inline protocol/jsonrpc/jsonrpc_invoker.go]: add log prefix "[JsonrpcInvoker][Invoke]"

File: common/utils/net.go (hunk 4 above)
Comment (lazyfighter) [reply common/utils/net.go]: why use `len`? i cant understand

File: common/utils/net.go (hunk 2 above)
Comment (lazyfighter) [reply common/utils/net.go]: ok i will run go fmt ./...

File: protocol/result.go [hunk 7]
```diff
@@ -18,13 +18,13 @@
 package protocol
 
 type Result interface {
-	SetError(error)
+	SetError(error) Result
```
Comment (fangyincheng) [inline protocol/result.go]: Why return Result? do you want  `SetError(e).SetResult(r)`? but you don't use this.？

**Pull request closed without merging**